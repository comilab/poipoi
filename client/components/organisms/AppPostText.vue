<template lang="pug">
div
  AppPostTextPage.my-4(
    v-for="(page, i) in pages",
    :key="i",
    :post="post",
    :page="page",
    :id="getPageId(i)",
    v-show="!showAllButtonIsVisible || i === 0",
    @jump="onJump"
  )
  .text-center.-mt-2(v-if="showAllButtonIsVisible")
    AppButton(
      color="white",
      :icon="faCopy",
      pill,
      @click="isAllShown = true"
    ) すべて表示 ({{ pages.length }}ページ)
</template>

<script lang="ts">
import { defineComponent, reactive, computed, toRefs } from 'nuxt-composition-api'
import VueScrollTo from 'vue-scrollto'
import { faCopy } from '@fortawesome/free-solid-svg-icons'

import Post from '~/models/Post'
import useUniqId from '~/composables/use-uniq-id'
import AppButton from '~/components/atoms/AppButton.vue'
import AppPostTextPage from '~/components/organisms/AppPostTextPage.vue'

export default defineComponent({
  components: {
    AppButton,
    AppPostTextPage
  },
  props: {
    post: {
      type: Post,
      required: true
    }
  },
  setup (props, { root }) {
    const { id } = useUniqId('post-text')

    const localState = reactive({
      isAllShown: false
    })

    const pages = computed(() => {
      return props.post.textPages
    })

    const showAllButtonIsVisible = computed(() => {
      return pages.value.length > 1 && !localState.isAllShown
    })

    const getPageId = (pageNum: number) => {
      return `${id}-${pageNum}`
    }

    const onJump = async (to: number) => {
      if (to > 0 && !localState.isAllShown) {
        localState.isAllShown = true
        await root.$nextTick()
      }
      VueScrollTo.scrollTo(`#${getPageId(to)}`)
    }

    return {
      ...toRefs(localState),
      pages,
      showAllButtonIsVisible,
      getPageId,
      onJump,
      faCopy
    }
  }
})
</script>
