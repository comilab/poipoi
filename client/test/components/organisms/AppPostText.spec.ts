import { fireEvent, render } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import VueScrollTo from 'vue-scrollto'

import AppPostText from '~/components/organisms/AppPostText.vue'
import postFactory from '~/test/factories/post'
import useUniqId from '~/composables/use-uniq-id'

jest.mock('vue-scrollto')
jest.mock('~/composables/use-uniq-id')

describe('components/organisms/AppPostText', () => {
  let options: any
  let uniqIdMock: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true,
        NuxtLink: true
      }
    }

    uniqIdMock = {
      id: 'foo'
    }
    mocked(useUniqId).mockReturnValue(uniqIdMock)
  })

  test('renders correctly', () => {
    options.props.post = postFactory.build({
      text: 'ページ1[newpage]ページ2'
    })
    const { container } = render(AppPostText, options)

    expect(container).toMatchSnapshot()
  })

  describe('すべて表示', () => {
    beforeEach(() => {
      options.props.post = postFactory.build({
        text: 'ページ1[newpage]ページ2'
      })
    })

    test('複数ページになる場合のみ表示', async () => {
      const { queryByRole, updateProps } = render(AppPostText, options)

      expect(queryByRole('button', { name: /すべて表示/ })).toBeInTheDocument()

      await updateProps({
        post: postFactory.build({
          text: 'ページ1'
        })
      })
      expect(queryByRole('button', { name: /すべて表示/ })).not.toBeInTheDocument()
    })

    test('ボタンを押すと全てのページを表示', async () => {
      const { container, getByRole } = render(AppPostText, options)

      expect(container.querySelector('#foo-0')).toBeVisible()
      expect(container.querySelector('#foo-1')).not.toBeVisible()

      await fireEvent.click(getByRole('button', { name: /すべて表示/ }))

      expect(container.querySelector('#foo-1')).toBeVisible()
    })
  })

  describe('ページジャンプ', () => {
    beforeEach(() => {
      options.props.post = postFactory.build({
        text: '[jump:2][newpage]ページ2'
      })
    })

    test('リンクを押すと指定されたページにジャンプ', async () => {
      const { getByText, container } = render(AppPostText, options)

      await fireEvent.click(getByText('2ページ目へ'))

      expect(VueScrollTo.scrollTo).toHaveBeenCalledWith('#foo-1')
      expect(container.querySelector('#foo-1')).toBeVisible()
    })
  })
})
