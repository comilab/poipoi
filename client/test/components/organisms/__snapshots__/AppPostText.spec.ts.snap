// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppPostText renders correctly 1`] = `
<div>
  <div>
    <section
      class="rounded bg-white my-8 p-4 my-4 text-page"
      data-testid="section"
      id="foo-0"
    >
      <!---->
      ページ1
    </section>
    <section
      class="rounded bg-white my-8 p-4 my-4 text-page"
      data-testid="section"
      id="foo-1"
      style="display: none;"
    >
      <!---->
      ページ2
    </section>
    <div
      class="text-center -mt-2"
    >
      <button
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded-full"
        data-testid="button"
        type="button"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw mr-2"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class=""
          data-testid="text"
        >
          すべて表示 (2ページ)
        </div>
      </button>
    </div>
  </div>
</div>
`;
